enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,244 @@
+8. Once the new CSV reaches the "Succeeded" phase the old CSV is removed from the cluster.
+
+The main reasons that an upgrade may fail include:
+- **Invalid CSV:** The CSV may come in the form of a failing CSV, or in cases of unmet requirements, a forever pending CSV.
+- **Invalid InstallPlan:** Usually occurs because a resource fails to be created or updated. An InstallPlan may fail independently of its CSV. The InstallPlan may fail even with a successfully applied CSV, or even a missing CSV.
+
+So, a failed workflow might look like:
+1. A User creates a subscription for a specific operator.
+2. OLM calculates the latest generation of operators that should be installed in a namespace and creates an InstallPlan listing each resource that must exist on the cluster.
+3. If the InstallPlan requires manual approval, a User must "approve" it.
+5. OLM creates/updates the manifests defined in the InstallPlan
+6. If an older CSV exists on the cluster OLM will place in the "Replacing" phase.
+7a. OLM fails to create/update a resource defined in the InstallPlan and moves the InstallPlan to the "Failed" state.
+7b. OLM installs a CSV and it enters the "Failed" state or forever enters the "Pending" state.
+
+In either failure scenario, a cluster admin must perform manual steps to recover, which becomes increasingly untenable as the number of managed clusters increases.
+
+> Note: OLM does not perform multiple attempts to resolve InstallPlans, meaning that a failed install blocks upgrades for all operators in the namespace as resolution is on a per-namespace basis.
+
+This enhancement proposes allowing cluster admins to opt-into fail forward upgrades, thereby enabling OLM to recover from failed installations when a new upgrade path is discovered.
+
+### User Stories
+
+- As a cluster admin, I want to allow all operators to "fail forward" in a specific namespace.
+- As a cluster admin, if an InstallPlan is in the FAILED state, I want to create the next InstallPlan.
+- As a cluster admin, if a CSV is stuck in the PENDING state, I want to allow it to upgrade to the next version.
+- As a cluster admin, if a CSV is stuck in the FAILED state, I want to allow it to upgrade to the next version.
+- As a catalog curator / operator author, I want to chose which versions of an operator may "fail forward" to the next version so my customers don't miss critical upgrades.
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Opting into Fail Forward Upgrades
+
+Since resolution is namespaced, the toggle for allowing forced upgrades also should be namespaced to avoid having to add the extra burden on the resolver surrounding understanding which dependent operators can or cannot use the forced upgrade toggle.
+
+With this in mind, this enhancement proposes adding the `failForwardUpgrades` toggle to an already watched namespaced resource, the OperatorGroup.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OperatorGroup
```

Comment:
I understand the motivation for using an OperatorGroup for the `failForwardUpgrade` toggle, but that API is viewed as all but deprecated and not in-use by the next generation of APIs, where operator tenancy is at the clusterwide level. Were any alternative ways to surface this toggle to the user considered? Might make sense to mention them. 

Also, since a missing OperatorGroup in a namespace is a legitimate reason for an install to fail, in that case failForward would not be possible (since the OperatorGroup is missing altogether). Might be worth calling out that edge-case.

Reply:
This could be as simple as a configmap, but this form of fail-forward is meant to be until we switch over to the new APIs. With the current resolver, the toggle needs to be namespaced since resolution is namespaced and OperatorGroups are a convenient place for that without having to manage one more resource.

In case of a missing OperatorGroup, once a new one is created with `failForwardUpgrade` set, the behavior would be identical to when it isn't.

Reply:
>I understand the motivation for using an OperatorGroup for the failForwardUpgrade toggle, but that API is viewed as all but deprecated and not in-use by the next generation of APIs

I am of the opinion that we're adding this feature to support existing workflows with OLM and handle it more gracefully in the next iteration of our APIs. As such, I want it to be on a resource that will allow us to pivot to a new solution as easily as possible in the next generation of APIs, which could include some form of migration logic.

>Also, since a missing OperatorGroup in a namespace is a legitimate reason for an install to fail, in that case failForward would not be possible (since the OperatorGroup is missing altogether). Might be worth calling out that edge-case.

Great point, failing forward would never be able to address this issue given the need for cluster admin to scope the operator so I'll be sure to mention this in the next update to the PR.